lib/actions/lookupObject.js

```diff
@@ -0,0 +1,108 @@
+const jsforce = require('jsforce');
+const debug = require('debug')('salesforce:lookup');
+const messages = require('elasticio-node').messages;
+const MetaLoader = require('../helpers/metaLoader');
+const SALESFORCE_API_VERSION = '45.0';
+
+/**
+ * This function will return a metamodel description for a particular object
+ *
+ * @param configuration
+ */
+module.exports.getMetaModel = async function getMetaModel(configuration) {
+  configuration.metaType = 'lookup';
+  const metaLoader = new MetaLoader(configuration);
+  const metaData = await metaLoader.loadMetadata();
+  metaData.in.properties[configuration.lookupField].required = !configuration.allowCriteriaToBeOmitted;
+  return metaData;
+};
+
+/**
+ * This function will return a metamodel description for a particular object
+ *
+ * @param configuration
+ */
+module.exports.getLookupFieldsModel = async function getLookupFieldsModel(configuration) {
+  const metaLoader = new MetaLoader(configuration);
+  return await metaLoader.getLookupFieldsModelWithTypeOfSearch(configuration.typeOfSearch);
+};
+
+/**
+ * See list on https://na45.salesforce.com/services/data/
+ * @type {string}
+ */
+module.exports.process = async function processAction(message, configuration) {
+  const { allowCriteriaToBeOmitted, allowZeroResults, lookupField } = configuration;
+  const lookupValue = message.body[lookupField];
+  let res = [];
+  const conn = new jsforce.Connection({
+    oauth2: {
+      clientId: process.env.SALESFORCE_KEY,
+      clientSecret: process.env.SALESFORCE_SECRET,
+      redirectUri: 'https://app.elastic.io/oauth/v2',
+    },
+    instanceUrl: configuration.oauth.instance_url,
+    accessToken: configuration.oauth.access_token,
+    refreshToken: configuration.oauth.refresh_token,
+    version: SALESFORCE_API_VERSION,
+  });
+
+  conn.on('refresh', (accessToken, res) => {
+    console.log('Keys were updated, res=%j', res);
+    this.emit('updateKeys', {oauth: res});
+  });
+
+  if (!lookupValue){
+    if(allowCriteriaToBeOmitted) {
+      this.emit('data', messages.newMessageWithBody({}));
+      return;
+    } else {
+      const err = new Error('No unique criteria provided');
+      console.error(err);
+      this.emit('error', err);
+      return;
+    }
+  }
+  await conn.sobject(configuration.sobject)
+      .select('*')
+      .where(
+          `${lookupField} = '${lookupValue}'`)
+      .on('record', (record) => {
+        res.push(record);
+      })
+      .on('end', () => {
+        if (!res.length) {
```

Comment:
I like `res.length === 0` but this is just preference.

Reply:
changed